### ISSUE TITLE
testing: Run the test suite against DNF nightlies

### ISSUE THREAD

**Issue by lukash** [osbuild/osbuild-composer#1011]
We have DNF nightly builds in a COPR repository:
https://copr.fedorainfracloud.org/coprs/rpmsoftwaremanagement/dnf-nightly/

It would be very useful to run osbuild-composer tests against these nightlies periodically (daily or with a lesser frequency) to catch any regressions early.

**Comment by teg** [comment]
Not sure the best way to achieve this, but I agree it would be a very nice thing to have. In the same way we want to test against `osbuild`.

**Comment by atodorov** [comment]
@lukash how do we know when the copr repository has been updated ? Is there a message or a property which we can use for triggering ?

**Comment by dvdhrm** [comment]
The `osbuild-composer` program pulls in `osbuild` and `dnf` as dependencies and relies on their stable API. Those programs are actively developed, as such their exposed API semantics as well as syntax might change (both deliberately and unintentionally).
So far, our arguing was that those dependencies run the `osbuild-composer` test-suite to verify they did not break API and their development did not unintentionally disrupt dependent projects. We already do this for `osbuild`, and I think it would be great to extend those capabilities and make our other dependencies do the same (in the same style as reverse-gating).

This testing style guarantees each project can test their *nightlies* against *stable* dependents/dependencies. The distribution gating mechanism serves as integration point if two parallel changes might cause hiccups.

Long story short: I think the way forward is to make the composer test-suite usable in other projects and let them provide custom repositories for the system setup and thus inject their nightlies into the test-suite.

**Comment by lukash** [comment]
> @lukash how do we know when the copr repository has been updated ? Is there a message or a property which we can use for triggering ?

I don't know about any mechanism in COPR that would allow this, maybe @praiskup would be able to quickly chime in...

In case there isn't any, it wouldn't be too bad to have a look at the times the builds usually finish, add some amount of time as padding and worst case scenario the tests will run against yesterday's build on the occasion COPR happens to be too slow. WDYT?

**Comment by atodorov** [comment]
> Long story short: I think the way forward is to make the composer test-suite usable in other projects and let them provide custom repositories for the system setup and thus inject their nightlies into the test-suite.

Look at this: https://src.osci.redhat.com/rpms/osbuild-composer/blob/rhel-8.4.0/f/tests/scripts/run_tests.sh - that's the closest we have to a 1-stop test execution for osbuild-composer. For the moment it is a subset of what happens in CI for every pull request here.  If you are looking for something more standardized then we're open for discussions and pull requests.

**Comment by atodorov** [comment]
> In case there isn't any, it wouldn't be too bad to have a look at the times the builds usually finish, add some amount of time as padding and worst case scenario the tests will run against yesterday's build on the occasion COPR happens to be too slow. WDYT?

We have the basic syntax to trigger via cron for RHEL nightly builds, still WIP but it's almost there - mainly blocked on other failing issues, not the pipeline configuration itself, see #976. Can adjust a bit for DNF nightlies and run them via cron too.

**Comment by lukash** [comment]
> The `osbuild-composer` program pulls in `osbuild` and `dnf` as dependencies and relies on their stable API. Those programs are actively developed, as such their exposed API semantics as well as syntax might change (both deliberately and unintentionally).
> So far, our arguing was that those dependencies run the `osbuild-composer` test-suite to verify they did not break API and their development did not unintentionally disrupt dependent projects. We already do this for `osbuild`, and I think it would be great to extend those capabilities and make our other dependencies do the same (in the same style as reverse-gating).
> 
> This testing style guarantees each project can test their _nightlies_ against _stable_ dependents/dependencies. The distribution gating mechanism serves as integration point if two parallel changes might cause hiccups.
> 
> Long story short: I think the way forward is to make the composer test-suite usable in other projects and let them provide custom repositories for the system setup and thus inject their nightlies into the test-suite.

I completely agree on conceptual level. We've had a discussion with @atodorov preceding this ticket, the first order of business was to attempt what you describe. However, the DNF CI currently supports only running test suites in containers and osbuild-composer tests require the creation of VMs, which we are not capable of.

DNF CI is right now running on a Red Hat internal Jenkins and is quite limited. We're not aware of a better infrastructure we could use for our CI and are also limited in dev power we can dedicate to this. Of course, a shared infrastructure where we could run dependant projects' CI in a standardized manner would be a dream :slightly_smiling_face:

As it stands, the VM requirement seems to be a dealbreaker for us now.

**Comment by lukash** [comment]
> We have the basic syntax to trigger via cron for RHEL nightly builds, still WIP but it's almost there - mainly blocked on other failing issues, not the pipeline configuration itself, see #976. Can adjust a bit for DNF nightlies and run them via cron too.

I think that's really good enough, unless there's an easy way to trigger from COPR, which I don't think there is.

**Comment by praiskup** [comment]
> I think that's really good enough, unless there's an easy way to trigger from COPR, which I don't think there is.

Copr sends [messages](https://apps.fedoraproject.org/datagrepper/raw?category=copr&delta=172800) to fedora-messaing bus, those messages can be (even anonymously) [consumed](https://copr-messaging.readthedocs.io/en/latest/).

**Comment by lukash** [comment]
Thanks, @praiskup, that's neat.

**Comment by atodorov** [comment]
> Copr sends [messages](https://apps.fedoraproject.org/datagrepper/raw?category=copr&delta=172800) to fedora-messaing bus, those messages can be (even anonymously) [consumed](https://copr-messaging.readthedocs.io/en/latest/).

@major can this be consumed easily by our pipeline, what do you think ?  From the docs above I see that we have to create some Python script and hook it up to execute as one of the first steps in a Pipeline and then conditionally decide whether or not there was a new DNF build. 

OTOH there's the Jenkins Copr plugin https://plugins.jenkins.io/copr/ but that has a very low usage count and IDK what syntax is supported for triggering pipeline execution when a new message arrives (if at all supported). There doesn't seem to be any documentation either.

**Comment by jrusz** [comment]
@atodorov In case this is still relevant, I belive this plugin could be used here for triggering https://plugins.jenkins.io/jms-messaging/ ... We should be able to configure the plugin to listen for messages on the fedora messaging bus and then configure a pipeline trigger to start a build when a specific message or topic is received...
